File: media/hs/src/Webar/Image/Source.hs
{-# LANGUAGE TemplateHaskell #-}

-- | generic image store objects
module Webar.Image.Source
  ( F.FetchInfo (..),
    FetchId,
    RequestId (..),
    Request,
    Response,
    HttpInfo (..),
    RequestRecord (..),
    Content (..),
    Image (..),
    SnapshotType (..),
  )
where

import Data.UUID.Types (UUID)
import Data.Void (Void)
import Webar.Blob
import Webar.Data.Cbor.TH (deriveProdCbor)
import Webar.Data.TH
import qualified Webar.Fetch.Http as F
import qualified Webar.Http as H
import Webar.Object (ObjectId)
import Webar.Types (Timestamp)

type FetchId = F.FetchId Void

newtype RequestId = XRequestId UUID
  deriving (Show, Eq)

deriveSumData
  SumOptions
    { sumProduct = ProductOptions {fieldLabelModifier = id},
      constructorTagModifier = camelTo2 '-'
    }
  ''RequestId

type Request = H.Request RequestId ()

type Response = H.Response () (BlobId ImageData)

data HttpInfo = HttpInfo
  { hiFetch :: FetchId,
    hiRequest :: Request,
    hiResponse :: Response
  }
  deriving (Show)

deriveProdCbor
  ProductOptions {fieldLabelModifier = camelTo2 '_' . drop 2}
  ''HttpInfo

data RequestRecord = RrFetch | RrHttpRequest
  deriving (Show, Eq)

deriveSumData
  SumOptions
    { sumProduct = ProductOptions {fieldLabelModifier = id},
      constructorTagModifier = camelTo2 '_' . drop 2
    }
  ''RequestRecord

newtype Content = CNormal (BlobId ImageData)
  deriving (Show)

deriveSumData
  SumOptions
    { sumProduct = ProductOptions {fieldLabelModifier = id},
      constructorTagModifier = camelTo2 '_' . tail
    }
  ''Content

data Image = Image
  { imgFetch :: FetchId,
    imgRequest :: ObjectId HttpInfo,
    imgTimestamp :: Timestamp,
    imgContent :: Content
  }
  deriving (Show)

deriveProdData
  ProductOptions {fieldLabelModifier = camelTo2 '_' . drop 3}
  ''Image

data SnapshotType = IsImage
  deriving (Show, Eq)

deriveSumData
  SumOptions
    { sumProduct = ProductOptions {fieldLabelModifier = id},
      constructorTagModifier = camelTo2 '_' . drop 2
    }
  ''SnapshotType

File: media/core-rs/src/image/source.rs
use serde::Deserialize;
use uuid::Uuid;

use webar_core::{
    blob::{BlobId, ImageData},
    fetch, http,
    object::ObjectId,
    Timestamp,
};
use webar_data::ser::{Never, Serialize};

pub type FetchInfo = fetch::http::FetchInfo<Never>;

pub type FetchId = fetch::http::FetchId<Never>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum RequestId {
    #[serde(rename = "x-request-id")]
    XRequestId(Uuid),
}

pub type Request<U> = http::Request<RequestId, U, (), ()>;
pub type Response<B> = http::Response<(), B>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HttpInfo<U> {
    pub fetch: FetchId,
    pub request: Request<U>,
    pub response: Response<BlobId<ImageData<Vec<u8>>>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum RequestRecord {
    #[serde(rename = "fetch")]
    Fetch,
    #[serde(rename = "http_request")]
    HttpRequest,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Content {
    #[serde(rename = "normal")]
    Normal(BlobId<ImageData<Vec<u8>>>),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Image {
    pub fetch: FetchId,
    pub request: ObjectId<HttpInfo<String>>,
    pub timestamp: Timestamp,
    pub content: Content,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum SnapshotType {
    #[serde(rename = "image")]
    Image,
}
